### PR TITLE
feat: Helpers to get the local interface MTU to a target IP address

### PR DESCRIPTION
**Pull request by larseggert**
WIP

### CODE REVIEW

File: neqo-transport/src/mtu.rs [hunk 1]
```diff
@@ -82,11 +82,12 @@ pub fn get_interface_mtu(remote_ip: &IpAddr) -> Result<u32, Error> {
         if let Some(iface) = iface {
             #[cfg(target_os = "macos")]
             {
-                // On macOS, We need to loop again to find the MTU of that interface. We need to do two
-                // loops, because `getifaddrs` returns one entry per interface and link type, and
-                // the IP addresses are in the AF_INET/AF_INET6 entries for an interface, whereas
-                // the MTU is (only) in the AF_LINK entry, whose `ifa_addr` contains MAC address
-                // information, not IP address information.
+                // On macOS, We need to loop again to find the MTU of that interface. We need to do
```
Comment (martinthomson) [inline neqo-transport/src/mtu.rs:85]:
```suggestion
                // On macOS, we need to loop again to find the MTU of that interface. We need to do
```

**Review by martinthomson — COMMENTED** [review]
Do we really want to run this sort of function once per connection?

Is there a design you can come up with that ensures that this only runs once?

**Comment by martinthomson** [comment]
I know that poking at the OS a few times per connection/path is not likely to cost that much, but it seems wasteful when in most cases the local interface MTU has no bearing on the end state.

**Comment by larseggert** [comment]
I mean, we could cache this value and only probe once in a while, but that would mean that if the cached value is too large, we PTO a bunch of times when opening the connection until we determine we should start PMTU from the lowest value. It's a tradeoff.

**Comment by martinthomson** [comment]
Why would we start with the assumption that the local MTU is the path MTU?  I have a 1500 byte MTU on my local link according to the OS, but I've never managed to get that much through.  Wouldn't we want to start low and then grow.  The local link MTU is likely a stopping point, not a starting point.

**Comment by larseggert** [comment]
Really, you don't get 1500 end to end (IP packet size)? What uplink do you have?

**Comment by larseggert** [comment]
TCP starts with the local MTU.

**Comment by larseggert** [comment]
So this seems to finally work also on Windows. 

I wonder if this should be in it's own crate, or maybe in `neqo-udp` if that might be intended to be a more generic crate?

After this has landed, I plan to do another PR to use this function to optimistically start with the interface MTU at the beginning of the handshake, and fall back to minimum-sized packets during the handshake followed bu regular PMTUD if our optimism was misguided.

**Review by mxinden — COMMENTED** [review]
:rocket: 

I will give this a review tomorrow. Though it will take some time given all the `unsafe`.

We might be able to make use of more static analysis tools like https://github.com/rust-lang/miri here.

I am in favor of making this a standalone crate published to crates.io. Given that it can be useful on its own, I don't think we should merge it with `neqo-udp`.

File: neqo-transport/src/mtu.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        0,
+    ))?;
+    socket.connect(remote)?;
+    let local_ip = socket.local_addr()?.ip();
+
+    assert_cfg!(any(
+        target_os = "macos",
+        target_os = "linux",
+        target_os = "windows"
+    ));
```
[the remark this answers is not in the document]
Comment (larseggert) [reply neqo-transport/src/mtu.rs:46]: I was trying to make it fail on non-supported platforms. Although on second thought, maybe just returning an error is enough here.

**Comment by larseggert** [comment]
> I am in favor of making this a standalone crate published to crates.io. Given that it can be useful on its own, I don't think we should merge it with `neqo-udp`.

Anyone got any ideas for a good crate name?

**Comment by mxinden** [comment]
> > I am in favor of making this a standalone crate published to crates.io. Given that it can be useful on its own, I don't think we should merge it with `neqo-udp`.
> 
> Anyone got any ideas for a good crate name?

`udp-mtu` or `mtu`?

https://crates.io/crates/mtu

**Comment by larseggert** [comment]
`mtu` is shortest :-)

**Comment by larseggert** [comment]
Closing, since this is now in https://github.com/mozilla/mtu

**Pull request closed without merging**